### ISSUE TITLE
Set up QL810WC label printer automation with AprilTags

### ISSUE THREAD

**Issue by sgbaird** [AccelerationConsortium/ac-training-lab#70]
https://github.com/AccelerationConsortium/echem-cell/issues/4

https://github.com/pklaus/brother_ql

https://stackoverflow.com/questions/55521301/printing-label-from-brother-ql-800-label-printer

i.e., with new pieces of equipment, autoprint an AprilTag to be affixed to it.

**Comment by sgbaird** [comment]
https://community.element14.com/products/raspberry-pi/raspberrypi_projects/b/blog/posts/use-a-raspberry-pi-zero-w-as-a-wireless-print-server

**Comment by sgbaird** [comment]
Parts arrived

![PXL_20241009_191126907.jpg](https://github.com/user-attachments/assets/cd85f882-6a0b-4972-a5a3-4fc934413ad9)

**Comment by SissiFeng** [comment]
Everything inside of the box
![image](https://github.com/user-attachments/assets/9da39e23-dc8b-432e-b66e-86a55fa87977)

**Comment by SissiFeng** [comment]
The printer failed to connect to the IoT WiFi (an email has been sent to support). 

The next step is to set up a Raspberry Pi connected to the label printer, configured for MQTT, and use MongoDB to store paired information (tag–device) and handle logging.

**Comment by SissiFeng** [comment]
The Pi is all set up, the printer driver is 32 bit and this model can't be installed via ppd file (maybe downgraded or otherwise universal driven). The app.py deployed to hugging face tests fine locally, just missing the printer connection part.

**Comment by sgbaird** [comment]
Ok, thanks for checking into that. That's unfortunate the driver doesn't work out of the box. Maybe we don't actually need the desktop driver and can instead use the Python package directly? I'm seeing a few threads related to this:

- https://www.rs-online.com/designspark/building-a-pi-powered-wireless-label-printer
- https://forums.raspberrypi.com/viewtopic.php?t=317783

**Comment by sgbaird** [comment]
@SissiFeng it sounds like these instructions (of course, switching the printer type that was mentioned) didn't work. https://www.rs-online.com/designspark/building-a-pi-powered-wireless-label-printer

Could you confirm?

**Comment by sgbaird** [comment]
What do you think about this? https://stackoverflow.com/a/70015794/13697228

**Comment by SissiFeng** [comment]
Brother QL-810W Printer Troubleshooting on Raspberry Pi Zero W

This summary outlines the key issues encountered and resolved so far, providing a reference for further troubleshooting.

---

1. System Architecture and Package Compatibility
- Brother's official drivers are for i386 (Intel) architecture, incompatible with Raspberry Pi’s ARM.
- Switching from a 64-bit to a 32-bit Raspberry Pi system solved some issues but required manual driver extraction.

2. Driver and Filter File Issues
- The PPD and filter files were extracted manually, but CUPS logs indicated the filter was “missing” despite being correctly placed.
- Potential symbolic link or permission issues affecting the filter.

3. Perl Script Execution
- The filter is a Perl script requiring Perl installation.
- Perl failed to run due to locale errors, which persisted even after generating the required locale.

4. Locale Configuration Problems
- The system lacked the `en_GB.UTF-8` locale, causing warnings during Perl execution.
- Locale generation reduced some warnings, but not all.

5. Unresponsive Print Jobs
- Print jobs are queued, but the printer remains idle.
- Jobs are stopped due to filter execution errors in CUPS.

6. TLS Connection Errors
- Repeated `TLS fatal alert` errors indicate potential encrypted communication issues between CUPS and the printer.

7. Hardware Detection and USB Communication
- The printer is recognized via `lsusb`, but communication problems persist, possibly due to configuration issues.

---

Unsolved Issues till now:

1. Filter Execution Error
  The system reports the filter as missing, even though it is correctly installed.

2. Printer Not Printing
  Jobs are sent and queued, but the printer stays idle, indicating a deeper communication or driver issue.

**Comment by SissiFeng** [comment]
Open an issue https://github.com/pklaus/brother_ql/issues/162

**Comment by sgbaird** [comment]
There are a few pieces here specific to wireless: https://support.brother.com/g/b/sp/faqlist.aspx?c=ca&lang=en&prod=lpql810weus&tabid=2

**Comment by SissiFeng** [comment]
> There are a few pieces here specific to wireless: https://support.brother.com/g/b/sp/faqlist.aspx?c=ca&lang=en&prod=lpql810weus&tabid=2

The wifi connection issue has been resolved.

**Comment by SissiFeng** [comment]
Create a wpa_supplicant.conf file and push the network configuration directly to the printer’s underlying system. This method is commonly used for IoT devices that support command-based serialization tools (such as lpadmin) to modify network credentials.

**Comment by SissiFeng** [comment]
The current issue is that while the MQTT-based printing workflow works locally, it fails to connect to the printer when deployed on Hugging Face (using gradio). 

The MQTT-based printing workflow involves a publisher running remotely, which sends print commands containing the image path and printer name. A subscriber on the local mac, receives the command and prints the corresponding AprilTag image from the specified local path.

**Comment by sgbaird** [comment]
Gradio wouldn't be connecting directly with the printer though right? I.e., A hugging face app would communicate with the Mac (or RPi), and the Mac would communicate with the printer.

Is it that you're unable to connect to the MQTT broker from the hugging face app?

**Comment by SissiFeng** [comment]
Yes. I tried setting up MQTT over websockets, also confirms QoS configuration consistency at publish and subscribe time, no luck.

**Comment by sgbaird** [comment]
I'll try to reproduce

**Comment by sgbaird** [comment]
Do you have a demo video of printing an AprilTag via Python on your computer you could share?

**Comment by SissiFeng** [comment]
https://github.com/user-attachments/assets/e89d7f6d-8520-4609-ad7f-8aa34160be10

**Comment by sgbaird** [comment]
@AccelerationConsortium/ac-training-lab have you experienced any issues with gradio not communicating well via MQTT / HiveMQ?

**Comment by gursi26** [comment]
not really, Gradio -> HiveMQ was pretty stable in everything i used it for. is there somewhere where i can see this code?

**Comment by sgbaird** [comment]
@SissiFeng can you provide links or snippets for the app and device code so @gursi26 can take a look?

**Comment by sgbaird** [comment]
Manual at https://support.brother.com/g/b/manualtop.aspx?c=ca&lang=en&prod=lpql810weus

**Comment by ZZZhaoz** [comment]
Hi Sterling,
Sorry for the delayed update on this project (Issue #70). The cobot has taken more time than expected.
I just wanted to confirm: is this project something that can be done fully online, or will it require frequent offline testing? I’d prefer to work on projects that can primarily be completed remotely. If there are projects like the previous apriltag demo notebook, where I was able to contribute more effectively online, I’d love to be involved in something similar.
That said, I’m happy to follow your direction and take on whatever is needed. I’d just love to be placed where I can contribute most effectively.
Best regards,
Jason

________________________________
发件人: Sterling G. Baird ***@***.***>
发送时间: 2025年6月14日 18:40
收件人: AccelerationConsortium/ac-training-lab ***@***.***>
抄送: Zihan Zhao ***@***.***>; Assign ***@***.***>
主题: Re: [AccelerationConsortium/ac-training-lab] Set up QL810WC label printer automation with AprilTags (Issue #70)


Assigned #70<https://github.com/AccelerationConsortium/ac-training-lab/issues/70> to @ZZZhaoz<https://github.com/ZZZhaoz>.

―
